Suggest matching commands if the user makes a typo

## Describe the Feature
<!-- Describe the requested Feature -->
Suggest matching commands if the user mistypes. For instance, `react-native-cli ini` should show up the help message along with a friendly suggestion such as `Did you mean init?`

## Possible Implementations
<!-- Describe how to implement the feature -->
We can make use of the [didyoumean.js](https://github.com/dcporter/didyoumean.js/) matching engine.

@thymikee it seems couple of type definition libraries are missing out. Installing the dependencies results in the following:-

```md
Building..................................................................packages/tools/src/groupFilesByType.ts:9:23 - error TS7016: Could not find a declaration file for module 'lodash'. '/home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/node_modules/lodash/lodash.js' implicitly has an 'any' type.
  Try `npm install @types/lodash` if it exists or add a new declaration (.d.ts) file containing `declare module 'lodash';`

9 import {groupBy} from 'lodash';
                        ~~~~~~~~

packages/tools/src/groupFilesByType.ts:10:18 - error TS2307: Cannot find module 'mime'.

10 import mime from 'mime';
                    ~~~~~~

packages/tools/src/groupFilesByType.ts:24:26 - error TS7006: Parameter 'type' implicitly has an 'any' type.

24   return groupBy(assets, type => (mime.getType(type) || '').split('/')[0]);
                            ~~~~

packages/debugger-ui/src/middleware/index.ts:1:25 - error TS2307: Cannot find module 'serve-static'.

1 import serveStatic from 'serve-static';
                          ~~~~~~~~~~~~~~


Found 4 errors.


Unable to build TypeScript definition files
Error: Command failed: node /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/node_modules/typescript/bin/tsc -b /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/packages/cli /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/packages/cli-types /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/packages/debugger-ui /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/packages/platform-android /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/packages/platform-ios /home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/packages/tools
    at makeError (/home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/node_modules/execa/index.js:174:9)
    at Function.module.exports.sync (/home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/node_modules/execa/index.js:338:15)
    at Object.<anonymous> (/home/james-the-hacker/CodeSpace/GitHub-Projects/react-native-cli/scripts/buildTs.js:38:9)
    at Module._compile (internal/modules/cjs/loader.js:868:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:879:10)
    at Module.load (internal/modules/cjs/loader.js:731:32)
    at Function.Module._load (internal/modules/cjs/loader.js:644:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:931:10)
    at internal/main/run_main_module.js:17:11
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ prebuild: `yarn build:ts`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ prebuild script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

--- a/packages/cli/package.json
+++ b/packages/cli/package.json
@@ -37,6 +37,7 @@
     "connect": "^3.6.5",
     "cosmiconfig": "^5.1.0",
     "deepmerge": "^3.2.0",
+    "didyoumean": "^1.2.1",
     "envinfo": "^7.1.0",
     "errorhandler": "^1.5.0",
     "execa": "^1.0.0",

--- a/packages/cli/src/index.ts
+++ b/packages/cli/src/index.ts
@@ -103,7 +103,7 @@ function printUnknownCommand(cmdName: string) {
 function suggestCommands(cmd: string) {
   const suggestion = didYouMean(cmd, commander.commands.map(cmd => cmd._name));
     if (suggestion) {
-  	  logger.warn(`\n\nDid you mean ${suggestion}?`)
+  	  logger.warn(`Did you mean ${suggestion}?`)
     }
 }
 

--- a/packages/cli/src/index.ts
+++ b/packages/cli/src/index.ts
@@ -22,10 +22,7 @@ commander
   .arguments('<command>')
   .action((cmd) => {
     printUnknownCommand(cmd);
-    const suggestion = didYouMean(cmd, commander.commands.map(cmd => cmd._name));
-    if (suggestion) {
-  	  logger.warn(`\n\nDid you mean ${suggestion}?`)
-    }
+    suggestCommands(cmd);
     process.exit(1);
   })
 
@@ -101,6 +98,13 @@ function printUnknownCommand(cmdName: string) {
   } else {
     commander.outputHelp();
   }
+}
+
+function suggestCommands(cmd: string) {
+  const suggestion = didYouMean(cmd, commander.commands.map(cmd => cmd._name));
+    if (suggestion) {
+  	  logger.warn(`\n\nDid you mean ${suggestion}?`)
+    }
 }
 
 /**

--- a/packages/cli/src/index.ts
+++ b/packages/cli/src/index.ts
@@ -90,7 +90,7 @@ function printUnknownCommand(cmdName: string) {
   const suggestion = didYouMean(cmdName, commander.commands.map(cmd => cmd._name));
   let errorMsg = `Unrecognized command "${chalk.bold(cmdName)}".`;
   if (suggestion) {
-    errorMsg += `Did you mean ${suggestion}?`;
+    errorMsg += ` Did you mean "${suggestion}"?`;
   }
   if (cmdName) {
     logger.error(errorMsg);

--- a/packages/cli/src/index.ts
+++ b/packages/cli/src/index.ts
@@ -1,6 +1,7 @@
 import chalk from 'chalk';
 import childProcess from 'child_process';
 import commander from 'commander';
+import didYouMean from 'didyoumean';
 import path from 'path';
 
 import {Command, Config} from '@react-native-community/cli-types';
@@ -17,10 +18,16 @@ commander
   .option('--version', 'Print CLI version')
   .option('--verbose', 'Increase logging verbosity');
 
-commander.on('command:*', () => {
-  printUnknownCommand(commander.args.join(' '));
-  process.exit(1);
-});
+commander
+  .arguments('<command>')
+  .action((cmd) => {
+    printUnknownCommand(cmd);
+    const suggestion = didYouMean(cmd, commander.commands.map(cmd => cmd._name));
+    if (suggestion) {
+  	  logger.warn(`\n\nDid you mean ${suggestion}?`)
+    }
+    process.exit(1);
+  })
 
 const handleError = (err: Error) => {
   if (commander.verbose) {

--- a/packages/cli/src/index.ts
+++ b/packages/cli/src/index.ts
@@ -22,7 +22,6 @@ commander
   .arguments('<command>')
   .action((cmd) => {
     printUnknownCommand(cmd);
-    suggestCommands(cmd);
     process.exit(1);
   })
 
@@ -88,8 +87,13 @@ function printHelpInformation(
 }
 
 function printUnknownCommand(cmdName: string) {
+  const suggestion = didYouMean(cmdName, commander.commands.map(cmd => cmd._name));
+  let errorMsg = `Unrecognized command "${chalk.bold(cmdName)}".`;
+  if (suggestion) {
+    errorMsg += `Did you mean ${suggestion}?`;
+  }
   if (cmdName) {
-    logger.error(`Unrecognized command "${chalk.bold(cmdName)}".`);
+    logger.error(errorMsg);
     logger.info(
       `Run ${chalk.bold(
         '"react-native --help"',
@@ -98,13 +102,6 @@ function printUnknownCommand(cmdName: string) {
   } else {
     commander.outputHelp();
   }
-}
-
-function suggestCommands(cmd: string) {
-  const suggestion = didYouMean(cmd, commander.commands.map(cmd => cmd._name));
-    if (suggestion) {
-  	  logger.warn(`Did you mean ${suggestion}?`)
-    }
 }
 
 /**

--- a/packages/cli/src/index.ts
+++ b/packages/cli/src/index.ts
@@ -18,12 +18,10 @@ commander
   .option('--version', 'Print CLI version')
   .option('--verbose', 'Increase logging verbosity');
 
-commander
-  .arguments('<command>')
-  .action((cmd) => {
-    printUnknownCommand(cmd);
-    process.exit(1);
-  })
+commander.arguments('<command>').action(cmd => {
+  printUnknownCommand(cmd);
+  process.exit(1);
+});
 
 const handleError = (err: Error) => {
   if (commander.verbose) {
@@ -87,7 +85,10 @@ function printHelpInformation(
 }
 
 function printUnknownCommand(cmdName: string) {
-  const suggestion = didYouMean(cmdName, commander.commands.map(cmd => cmd._name));
+  const suggestion = didYouMean(
+    cmdName,
+    commander.commands.map(cmd => cmd._name),
+  );
   let errorMsg = `Unrecognized command "${chalk.bold(cmdName)}".`;
   if (suggestion) {
     errorMsg += ` Did you mean "${suggestion}"?`;

